x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/rule/ParameterizedRuleExecutor.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ql.rule;
+
+import org.elasticsearch.xpack.ql.tree.Node;
+
+import java.util.function.Function;
+
+public abstract class ParameterizedRuleExecutor<TreeType extends Node<TreeType>, Context> extends RuleExecutor<TreeType> {
+
+    private final Context context;
+
+    protected ParameterizedRuleExecutor(Context context) {
+        this.context = context;
+    }
+
+    protected Context context() {
+        return context;
+    }
+
+    @Override
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    protected Function<TreeType, TreeType> transform(Rule<?, TreeType> rule) {
```

Comment:
I agree with you it's not ideal - I picked this compromise to minimize the amount of disruption and see what the team thinks about it.
The issue is not the piping but also retrofitting the rule executor - if there's no state stored inside the instance then the `execute(TreeType)` method needs to become `execute(TreeType, Context)` with the later being invalid since one cannot execute a rule with a context.
A potential future PR would be to combine the Rule and RuleExecutor into one so there's always execution with context - when none is necessary `Void` can be used or a rule can simply ignore the context.